okhttp/src/main/java/io/grpc/okhttp/OkHttpFrameLogger.java

```diff
@@ -0,0 +1,162 @@
+package io.grpc.okhttp;
+
+import io.grpc.okhttp.internal.framed.ErrorCode;
+import io.grpc.okhttp.internal.framed.Header;
+import io.grpc.okhttp.internal.framed.Settings;
+import okio.Buffer;
+import okio.ByteString;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+public class OkHttpFrameLogger {
+    public enum Direction {
+        INBOUND,
+        OUTBOUND
+    }
+
+    private static final int BUFFER_LENGTH_THRESHOLD = 64;
+    private final Logger logger;
+    private final Level level;
+
+    public OkHttpFrameLogger(Level level) {
+        this(level, Logger.getLogger(OkHttpFrameLogger.class.getName()));
+    }
+
+    public OkHttpFrameLogger(Level level, String name) {
+        this(level, Logger.getLogger(name));
+    }
+
+    public OkHttpFrameLogger(Level level, Class<?> clazz) {
+        this(level, Logger.getLogger(clazz.getName()));
+    }
+
+    private OkHttpFrameLogger(Level level, Logger logger) {
+        this.level = checkNotNull(level, "level");
+        this.logger = checkNotNull(logger, "logger");
+    }
+
+    public boolean isEnabled() {
+        return logger.isLoggable(level);
+    }
+
+    public void logData(Direction direction, int streamId, Buffer data, int length, boolean endStream) {
+        if (isEnabled()) {
+            logger.log(level, direction.name() + " DATA: streamId=" + streamId
+                    + " endStream=" + endStream + " length=" + length + " bytes=" + data.snapshot(length));
+        }
+    }
+
+    public void logHeaders(Direction direction, int streamId, List<Header> headers, boolean endStream) {
+        if (isEnabled()) {
+            logger.log(level, direction.name() + " HEADERS: streamId=" + streamId + " headers="
```

Comment:
Use `direction +`, which will call `direction.toString()`, which is flexible to be overridden.